libs/zedUpload/httputil/http.go

```diff
@@ -148,30 +150,112 @@ func ExecCmd(cmd, host, remoteFile, localFile string, objSize int64,
 			return stats
 		}
 		defer local.Close()
-		defer resp.Body.Close()
-		chunkSize := SingleMB
-		var written, copiedSize int64
-		stats.Size = objSize
-		for {
-			var copyErr error
-			if written, copyErr = io.CopyN(local, resp.Body, chunkSize); copyErr != nil && copyErr != io.EOF {
-				stats.Error = copyErr
-				return stats
+
+		var errorList []string
+		done := false
+		canSeek := false
+
+		delay := time.Second
+		lastModified := ""
+		var currentErrorList []string
+		for attempt := 0; attempt < maxRetries; attempt++ {
+			if len(currentErrorList) > 0 {
+				// if we have previous errors, collect them
+				errorList = append(errorList, fmt.Sprintf("(attempt %d/%d):", attempt, maxRetries))
+				errorList = append(errorList, strings.Join(currentErrorList, " "))
+				currentErrorList = []string{}
 			}
-			copiedSize += written
-			if written != chunkSize {
-				// Must have reached EOF
+			if ctx.Err() != nil {
+				errorList = append(errorList, fmt.Sprintf("ctx.Err: %v", ctx.Err()))
 				break
 			}
-			stats.Asize = copiedSize
-			if prgNotify != nil {
-				select {
-				case prgNotify <- stats:
-				default: //ignore we cannot write
+			if done {
+				break
+			}
+			if attempt > 0 {
+				time.Sleep(delay)
+				if delay < maxDelay {
+					delay = delay * 2
+				}
+			}
+			if canSeek && copiedSize > 0 {
+				req.Header.Set("Range", fmt.Sprintf("bytes=%d-", copiedSize))
+			}
+			resp, err := client.Do(req)
+			if err != nil {
+				currentErrorList = append(currentErrorList, fmt.Sprintf("get failed: %s", err))
+				continue
+			}
+			if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusPartialContent {
+				currentErrorList = append(currentErrorList, fmt.Sprintf("bad response code: %d", resp.StatusCode))
+				err = resp.Body.Close()
+				if err != nil {
+					currentErrorList = append(currentErrorList, fmt.Sprintf("error close Body: %v", err))
+				}
+				errorList = append(errorList, strings.Join(currentErrorList, " "))
+				//we do not want to process server errors here
+				break
+			}
+			newLastModified := resp.Header.Get("Last-Modified")
+			if lastModified != "" && newLastModified != lastModified {
+				// last modified changed, retry without Range
```

Comment:
In this case the intent is to start over since the content changed, right?
If so don't you gave to set copiedSize = 0?
What about content already written to the file? Don't you need local.Truncate(0) as well?

Reply:
In this case we will continue attempt loop with canSeek=false, which will not fire `req.Header.Set("Range", fmt.Sprintf("bytes=%d-", copiedSize))`, so, server will response with `http.StatusOK` (not `http.StatusPartialContent`, which is response for request with Range in headers), so, we will go into the `if copiedSize > 0 && (!canSeek || resp.StatusCode != http.StatusPartialContent) {` below and will truncate file.

Reply:
But you already wrote something to the file, right? Some previous loop will have done that and then in the second or fourteenth loop interation you get the change to last-modified.

Reply:
Added `forceRestart` variable to schedule truncate on the next iteration

Reply:
OK